### PR TITLE
Modernize

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
Actions:

- GitHub 
- Actions instead of Travis CI
- EDoc readiness
- no shared `.hrl` files
- static analysis with `hank`
- static analysis with `elvis`
- broader static analysis with Xref
- multiple `-spec(_).` fixes
- simplified Docker-based tests approach
- minor `.y[a]ml`/`.md` tweaks
- modernized `Makefile`
- min. OTP bumped to 22 (from 19.1)
- covered OTP 24
- tweaked `rebar.config`

**Edit**: this action should be further publishable under `mc_gen_rpc`.
**Edit**: this pull request completely covers 2.1.2 while adding whatever was in `master` most recently; 2.2.0 (awkward, but for continuation) can be cut from `master` after this is accepted.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,53 +1,28 @@
-# gen_rpc: A scalable RPC library for Erlang-VM based languages
+# gen_rpc: a scalable RPC library for BEAM (Erlang-VM) based languages [![GitHub Actions CI][ci-img]][ci]
+
+[ci]: https://github.com/miniclip/gen_rpc
+[ci-img]: https://github.com/miniclip/gen_rpc/workflows/build/badge.svg
 
 ## Overview
 
-- Latest release: ![Tag Version](https://img.shields.io/github/tag/priestjim/gen_rpc.svg)
```
Comment (paulo-ferraz-oliveira) [inline README.md:5]: Some stuff moved to `miniclip`.
Status per branch removed, since it's on the top, now (and only for `master`)

File: README.md [hunk 2]
```diff
@@ -1,53 +1,28 @@
...
-- Branch status (`develop`): [![Build Status](https://travis-ci.org/priestjim/gen_rpc.svg?branch=develop)](https://travis-ci.org/priestjim/gen_rpc) [![Coverage Status](https://coveralls.io/repos/priestjim/gen_rpc/badge.svg?branch=develop&service=github)](https://coveralls.io/github/priestjim/gen_rpc?branch=develop)
-- Issues: [![GitHub issues](https://img.shields.io/github/issues/priestjim/gen_rpc.svg)](https://github.com/priestjim/gen_rpc/issues)
-- License: [![GitHub license](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/priestjim/gen_rpc/master/LICENSE)
+- Latest release: ![Tag Version](https://img.shields.io/github/tag/miniclip/gen_rpc.svg)
+- Issues: [![GitHub issues](https://img.shields.io/github/issues/miniclip/gen_rpc.svg)](https://github.com/miniclip/gen_rpc/issues)
+- License: [![GitHub license](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/miniclip/gen_rpc/master/LICENSE)
 - [Erlang Factory 2016 Talk](https://www.youtube.com/watch?feature=player_embedded&v=xiPnLACtNeo)
 
-## Build Dependencies
```
Comment (paulo-ferraz-oliveira) [inline README.md:12]: Removed since not really interesting.

File: include/guards.hrl [hunk 3]
```diff
@@ -1,13 +0,0 @@
-%%% -*-mode:erlang;coding:utf-8;tab-width:4;c-basic-offset:4;indent-tabs-mode:()-*-
```
Comment (paulo-ferraz-oliveira) [inline include/guards.hrl:1]: Not deleted, but moved and then changed: https://github.com/miniclip/gen_rpc/pull/4/files#diff-f8110bb2365868bb428d41358cb158bf6297756944dd57de101affa5be85eb1b.

File: include/ssl.hrl [hunk 4]
```diff
@@ -1,45 +0,0 @@
-%%% -*-mode:erlang;coding:utf-8;tab-width:4;c-basic-offset:4;indent-tabs-mode:()-*-
```
Comment (paulo-ferraz-oliveira) [inline include/ssl.hrl:1]: Deleted. All the stuff from this file was used by a single `.erl`.

File: package.exs [hunk 5]
```diff
@@ -3,16 +3,16 @@ defmodule GenRPC.Mixfile do
...
     [app: :gen_rpc,
-     version: "2.0.0",
+     version: "2.1.0",
      description: "A scalable RPC library for Erlang-VM based languages",
      package: package]
   end
 
   defp package do
     [files: ~w(include src LICENSE Makefile package.exs README.md TODO.md CHANGELOG.md rebar.config rebar.config.script otp-release.escript),
-     maintainers: ["Panagiotis PJ Papadomitsos"],
```
Comment (paulo-ferraz-oliveira) [inline package.exs:13]: I think this should be Ok in the context of our maintenance.

File: package.exs [hunk 6]
```diff
@@ -3,16 +3,16 @@ defmodule GenRPC.Mixfile do
...
+     maintainers: ["Miniclip"],
      licenses: ["Apache 2.0"],
-     links: %{"GitHub" => "https://github.com/priestjim/gen_rpc"}]
```
Comment (paulo-ferraz-oliveira) [inline package.exs:15]: I think this should be Ok in the context of our maintenance.

File: rebar.config [hunk 7]
```diff
@@ -1,106 +1,125 @@
...
+{minimum_otp_vsn, "22"}.
 
-%% Plugins
-{plugins, [rebar3_hex]}.
-{shell, [
-    {apps, [lager, sync, gen_rpc]}
-]}.
-
-{erl_opts, [debug_info,
-    {warn_format, 1},
```
Comment (paulo-ferraz-oliveira) [inline rebar.config:15]: Most of this is moving stuff around, but some (default) options were actually removed to save on maintenance costs.

File: rebar.config.script [hunk 8]
```diff
@@ -1,14 +0,0 @@
-%%% -*-mode:erlang;coding:utf-8;tab-width:4;c-basic-offset:4;indent-tabs-mode:()-*-
```
Comment (paulo-ferraz-oliveira) [inline rebar.config.script:1]: Not using `coveralls` any more.

File: src/gen_rpc.app.src [hunk 9]
```diff
@@ -9,9 +9,9 @@
     {vsn, git},
     {mod, {gen_rpc_app, []}},
     {registered, [gen_rpc_dispatcher]},
-    {maintainers, ["Panagiotis PJ Papadomitsos"]},
```
Comment (paulo-ferraz-oliveira) [inline src/gen_rpc.app.src:12]: I think this should be Ok in the context of our maintenance.

File: src/gen_rpc.app.src [hunk 10]
```diff
@@ -9,9 +9,9 @@
...
+    {maintainers, ["Miniclip"]},
     {licenses, ["Apache 2.0"]},
-    {links, [{"Github", "https://github.com/priestjim/gen_rpc"}]},
```
Comment (paulo-ferraz-oliveira) [inline src/gen_rpc.app.src:14]: I think this should be Ok in the context of our maintenance.

File: src/gen_rpc_client.erl [hunk 11]
```diff
@@ -66,27 +74,27 @@ stop(NodeOrTuple) when ?is_node_or_tuple(NodeOrTuple) ->
...
 call(NodeOrTuple, M, F, A, RecvTO) ->
     call(NodeOrTuple, M, F, A, RecvTO, undefined).
 
 %% Simple server call with custom receive and send timeout values
 %% This is the function that all of the above call
--spec call(node_or_tuple(), atom() | tuple(), atom() | function(), list(), timeout() | undefined, timeout() | undefined) ->
+-spec call(node_or_tuple(), atom() | tuple(), atom(), list(), timeout() | undefined, timeout() | undefined) ->
     term() | {badrpc,term()} | {badtcp,term()}.
 call(NodeOrTuple, M, F, A, RecvTO, SendTO) when ?is_node_or_tuple(NodeOrTuple), is_atom(M) orelse is_tuple(M), is_atom(F), is_list(A),
-                                         RecvTO =:= undefined orelse ?is_timeout(RecvTO),
```
Comment (paulo-ferraz-oliveira) [inline src/gen_rpc_client.erl:88]: Let me know if you prefer `is_timeout` (as was).

File: test/ct/remote_with_key_SUITE.erl [hunk 12]
```diff
@@ -143,7 +142,7 @@ random_local_tcp_close(_Config) ->
     true = erlang:exit(AccPid, kill),
     ok = timer:sleep(600), % Give some time to the supervisor to kill the children
     [?MASTER] = rpc:call(?SLAVE, gen_rpc, nodes, []),
-    [_Other] = supervisor:which_children(gen_rpc_acceptor_sup),
+    [_Other2] = supervisor:which_children(gen_rpc_acceptor_sup),
```
Comment (paulo-ferraz-oliveira) [inline test/ct/remote_with_key_SUITE.erl:146]: Compiler warning.

File: test/ct/remote_with_key_SUITE.erl [hunk 13]
```diff
@@ -154,7 +153,7 @@ random_remote_tcp_close(_Config) ->
     ok = timer:sleep(600),
     [?SLAVE] = gen_rpc:nodes(),
     [_Client] = supervisor:which_children(gen_rpc_client_sup),
-    [_Other] = rpc:call(?SLAVE, supervisor, which_children, [gen_rpc_acceptor_sup]).
+    [_Other2] = rpc:call(?SLAVE, supervisor, which_children, [gen_rpc_acceptor_sup]).
```
Comment (paulo-ferraz-oliveira) [inline test/ct/remote_with_key_SUITE.erl:157]: Compiler warning.

File: test/gen_rpc.coverspec [hunk 14]
```diff
@@ -1,5 +0,0 @@
-{nodes, ['gen_rpc_master@127.0.0.1', 'gen_rpc_slave1@127.0.0.1']}.
```
Comment (paulo-ferraz-oliveira) [inline test/gen_rpc.coverspec:1]: `rebar3` doesn't support cover spec.s (at least that's what the UI says). At the same time we're not exposing this info anywhere and there are more modern approaches to this problem.

File: test/integration/.dockerignore [hunk 15]
```diff
@@ -1,26 +1,5 @@
 .rebar3
```
Comment (paulo-ferraz-oliveira) [inline test/integration/.dockerignore:1]: The same as `.gitignore`.

File: test/integration/Dockerfile [hunk 16]
```diff
@@ -1,14 +1,5 @@
 # Dockerfile to test gen_rpc via proper distributed nodes
```
Comment (paulo-ferraz-oliveira) [inline test/integration/Dockerfile:1]: Simplified, simply. Works with `make integration` if you're curious.

File: test/integration/integration-tests.sh [hunk 17]
```diff
@@ -10,40 +10,27 @@ NODES=""
...
+	docker exec -it ${NAME} bash -c 'cd /gen_rpc && rebar3 as dev compile'
+    if [ ${1} != "master" ] ; then
+        docker exec -itd ${NAME} bash -c "cd /gen_rpc && rebar3 as dev shell --name gen_rpc@${IP}"
+    else
+        docker exec -it gen_rpc_master bash -c "export NODES=${NODES} NODE=gen_rpc@${IP} && cd /gen_rpc && rebar3 as dev compile && rebar3 ct --suite test/ct/integration_SUITE"
+    fi
 	return $?
 }
 
-start_master() {
```
Comment (paulo-ferraz-oliveira) [inline test/integration/integration-tests.sh:29]: This big block was _merged_ into the previous one.

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,6 +1,9 @@
 # Changelog
 
-Below is a non-exhaustive list of changes between `gen_rpc` versions.
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
```
Comment (g-andrade) [inline CHANGELOG.md]: ^ Should it be mentioned that the project follows Semantic Versioning _starting_ at whatever version it did? (Perhaps the next one to be released?)

I.e. "This project adheres to [Semantic Versioning](...) starting at version X.Y.Z"

File: Makefile [hunk 19]
```diff
@@ -1,148 +1,99 @@
-#
 # Copyright 2015 Panagiotis Papadomitsos. All Rights Reserved.
```
Comment (g-andrade) [inline Makefile:1]: ^ Should Miniclip's own copyright notice be placed below?

Otherwise it sort of implies that our fellow Panagiotis here made all the stuff we did, down below - I don't know if that might present a legal problem.

But IANAL.

File: README.md (hunk 1 above)
Comment (g-andrade) [reply README.md:5]: 👍

File: README.md (hunk 2 above)
Comment (g-andrade) [reply README.md:12]: 👍

File: README.md [hunk 20]
```diff
@@ -318,6 +311,6 @@ This project is published and distributed under the [Apache License](LICENSE).
 
 Please see [CONTRIBUTING.md](CONTRIBUTING.md)
 
-### Contributors:
+### Contributors
 
 - [Edward Tsang](https://github.com/linearregression)
```
Comment (g-andrade) [inline README.md]: ^ Should we collectively or individually self-aggrandize, here?

Or rather, my question is:
- should we keep the list and therefore keep it up-to-date or
- remove it so that we don't have to update it?

(It's the in-between approach that makes me uneasy)

File: package.exs (hunk 5 above)
Comment (g-andrade) [reply package.exs:13]: ✅

File: package.exs [hunk 21]
```diff
@@ -3,16 +3,16 @@ defmodule GenRPC.Mixfile do
```
Comment (g-andrade) [inline package.exs]: Concerning `package.exs` - we should just drop it.  It's an old way of publishing Erlang libraries to Hex, requiring Mix.

But I'll leave it up to you whether to handle this now or later.

File: rebar.config [hunk 22]
```diff
@@ -1,106 +1,125 @@
...
+    {i, "src"}
 ]}.
 
+%% == Dependencies and plugins ==
+
 {deps, [
-    {hut, "~> 1.2"},
-    {ssl_verify_fun, "~> 1.1"}
+    {hut, "1.3.0"},
+    {ssl_verify_hostname, "1.0.6"}
```
Comment (g-andrade) [inline rebar.config]: 🤔 

Why the change from `ssl_verify_fun` to `ssl_verify_hostname`?

Comment (g-andrade) [reply rebar.config]: `ssl_verify_hostname`'s GitHub link in Hex points to `ssl_verify_fun`'s repo 😵‍💫

File: CHANGELOG.md (hunk 18 above)
Comment (paulo-ferraz-oliveira) [reply CHANGELOG.md]: Nice catch. Not sure about the past, but we can play on the safe side.

File: Makefile (hunk 19 above)
Comment (paulo-ferraz-oliveira) [reply Makefile:1]: We've not done it before, but probably can. I'll put this in the files I changed.

File: rebar.config (hunk 22 above)
Comment (g-andrade) [reply rebar.config]: On Hex: the last version of `ssl_verify_hostname` is `1.0.6`, the first version of `ssl_verify_fun` is `1.0.8`, so I'm gonna throw a guess that the library was renamed.

If my hypothesis is true, it implies we're actually importing an older version of `ssl_verify_fun` (pre-rename) and should therefore keep using `ssl_verify_fun`.

File: CHANGELOG.md (hunk 18 above)
Comment (paulo-ferraz-oliveira) [reply CHANGELOG.md]: 6aee28d

File: src/driver/gen_rpc_driver_ssl.erl [hunk 23]
```diff
@@ -37,6 +35,46 @@
...
+                  "AES256-GCM-SHA384","AES256-SHA256","ECDHE-ECDSA-AES128-GCM-SHA256",
+                  "ECDHE-RSA-AES128-GCM-SHA256","ECDHE-ECDSA-AES128-SHA256","ECDHE-RSA-AES128-SHA256",
+                  "ECDH-ECDSA-AES128-GCM-SHA256","ECDH-RSA-AES128-GCM-SHA256","ECDH-ECDSA-AES128-SHA256",
+                  "ECDH-RSA-AES128-SHA256","DHE-DSS-AES128-GCM-SHA256","DHE-DSS-AES128-SHA256","AES128-GCM-SHA256",
+                  "AES128-SHA256","ECDHE-ECDSA-AES256-SHA","ECDHE-RSA-AES256-SHA","DHE-DSS-AES256-SHA",
+                  "ECDH-ECDSA-AES256-SHA","ECDH-RSA-AES256-SHA","AES256-SHA","ECDHE-ECDSA-AES128-SHA",
+                  "ECDHE-RSA-AES128-SHA","DHE-DSS-AES128-SHA","ECDH-ECDSA-AES128-SHA","ECDH-RSA-AES128-SHA","AES128-SHA"]},
+        {secure_renegotiate,true},
+        {reuse_sessions,true},
+        {versions,['tlsv1.2','tlsv1.1']},
```
Comment (g-andrade) [inline src/driver/gen_rpc_driver_ssl.erl]: ^ TLS v1.1? Mmmmh...

The list of `ciphers` also smells - these things tend to get outdated quickly when unmaintained, allowing for the use of vulnerable ciphers.

I'll need to investigate what's our best option here. Either
- drop both `ciphers` and `versions` and leave it to OTP to decide (might this risk insecure setups?)
- or check what are the latest recommendations from some reliable authority (e.g. Mozilla) and change `ciphers` and `versions` accordingly

File: src/driver/gen_rpc_driver_ssl.erl [hunk 24]
```diff
@@ -62,17 +100,17 @@ listen(Port) when is_integer(Port) ->
...
--spec accept(ssl:sslsocket()) -> ok | {error, term()}.
+-spec accept(ssl:sslsocket()) -> {ok, ssl:sslsocket()} | {error, term()}.
 accept(Socket) when is_tuple(Socket) ->
     {ok, TSocket} = ssl:transport_accept(Socket, infinity),
-    case ssl:ssl_accept(TSocket) of
-        ok ->
-            {ok, TSocket};
+    case ssl:handshake(TSocket) of
+        {ok, SSocket} ->
+            {ok, SSocket};
```
Comment (g-andrade) [inline src/driver/gen_rpc_driver_ssl.erl:114]: I'd replace the `ssl:transport_accept/2` + `ssl:handshake/1` calls with a single `ssl:accept/2` - it would definitely count as modernizing.

Also, `ssl:transport_accept/2` might return `{error, _}`s.

But I'll leave it up to you.

Comment (g-andrade) [reply src/driver/gen_rpc_driver_ssl.erl:114]: Wait, there is no such thing (`ssl:accept/2`.)

Where did this idea come from?

Oh well. Nothing to see, here.

Comment (g-andrade) [reply src/driver/gen_rpc_driver_ssl.erl:114]: (I can't resolve my own conversations, apparently)

File: Makefile (hunk 19 above)
Comment (paulo-ferraz-oliveira) [reply Makefile:1]: 801d0f4

File: src/gen_rpc_client.erl (hunk 11 above)
Comment (g-andrade) [reply src/gen_rpc_client.erl:88]: Nah, whatever there was in me that cared about the finer points of macro-presented-like-functions has dissolved into benevolent indifference.

File: test/integration/Dockerfile (hunk 16 above)
Comment (g-andrade) [reply test/integration/Dockerfile:1]: 👍

**Review by g-andrade — CHANGES_REQUESTED** [review]

File: test/gen_rpc.coverspec (hunk 14 above)
Comment (g-andrade) [reply test/gen_rpc.coverspec:1]: 👍

File: CHANGELOG.md (hunk 18 above)
Comment (g-andrade) [reply CHANGELOG.md]: 👍

File: Makefile (hunk 19 above)
Comment (g-andrade) [reply Makefile:1]: 👍

File: src/driver/gen_rpc_driver_ssl.erl (hunk 24 above)
Comment (g-andrade) [reply src/driver/gen_rpc_driver_ssl.erl:114]: (Ah, actually I can, but only after I finish a review.)

File: src/gen_rpc_client.erl (hunk 11 above)
Comment (g-andrade) [reply src/gen_rpc_client.erl:88]: (But thanks for asking)

File: README.md (hunk 20 above)
Comment (paulo-ferraz-oliveira) [reply README.md]: Yeah, I thought about that. I would say we drop it. On GitHub, at least, you always have the full list of contributors.

Comment (paulo-ferraz-oliveira) [reply README.md]: 8026891

File: package.exs (hunk 21 above)
Comment (paulo-ferraz-oliveira) [reply package.exs]: I'll do it now.

File: rebar.config (hunk 22 above)
Comment (paulo-ferraz-oliveira) [reply rebar.config]: The code uses `ssl_verify_hostname`, so I'm not actually sure what happened there. We might want to move to `ssl_verify_fun`, sure. `ssl_verify_hostname` is `ssl_verify_fun` renamed. Not sure about the versions, since I didn't check. Will do, later.

File: src/driver/gen_rpc_driver_ssl.erl (hunk 23 above)
Comment (paulo-ferraz-oliveira) [reply src/driver/gen_rpc_driver_ssl.erl]: Yeah, I just moved code around, here; didn't try to innovate. You might want to do that, but I propose doing it in a different pull request (you might want to create a task, for starters).

Comment (g-andrade) [reply src/driver/gen_rpc_driver_ssl.erl]: Wouldn't it count as _modernizing_ (and therefore be part of this PR)? 🙂 

(That is: updating outdated stuff.)

File: rebar.config (hunk 22 above)
Comment (g-andrade) [reply rebar.config]: `ssl_verify_fun` includes a module named `ssl_verify_hostname`, that's what likely misguided you

Comment (paulo-ferraz-oliveira) [reply rebar.config]: 👍 let me check this.

Comment (paulo-ferraz-oliveira) [reply rebar.config]:
```
src/driver/gen_rpc_driver_ssl.erl
Line 181 Column 49: Unknown function ssl_verify_hostname:verify_cert_hostname/2
Line 256 Column 33: Unknown function ssl_verify_hostname:verify_fun/3
```

(will dig deeper)

File: src/driver/gen_rpc_driver_ssl.erl (hunk 23 above)
Comment (paulo-ferraz-oliveira) [reply src/driver/gen_rpc_driver_ssl.erl]: > Wouldn't it count as _modernizing_ 

Not for me 😄. It's working. You can push on top of my changes once I'm done, though, but I don't want to put everything into a single pull request: it increases the chances of "never-endingness".

Comment (paulo-ferraz-oliveira) [reply src/driver/gen_rpc_driver_ssl.erl]: (I will, however, dedicate some minutes to this - if I'm unable to find the info. I'm looking for, I'll leave this juice for you)

File: rebar.config (hunk 22 above)
Comment (paulo-ferraz-oliveira) [reply rebar.config]: I think this mess is now untangled with 57adedb. `test` was pulling in a different lib./reference and screwing the analysis.

File: src/driver/gen_rpc_driver_ssl.erl (hunk 23 above)
Comment (paulo-ferraz-oliveira) [reply src/driver/gen_rpc_driver_ssl.erl]: There you go: 044fb46.

Source: https://wiki.mozilla.org/Security/Server_Side_TLS

File: rebar.config (hunk 22 above)
Comment (g-andrade) [reply rebar.config]: 😵 

👍

File: src/driver/gen_rpc_driver_ssl.erl (hunk 23 above)
Comment (g-andrade) [reply src/driver/gen_rpc_driver_ssl.erl]: Very nice, much secure

**Comment by g-andrade** [comment]
⚠️ Should `CHANGELOG.md` be updated?

(I didn't really consider this before, but the TLS-related  changes made me think about it - e.g. the consumer upgrades `gen_rpc`, runs into issues because of it [TLS] and doesn't know where to start.

[This, I guess, is where the thin line between modernizing and _modernizing_ stands.]

I don't mind updating `CHANGELOG.md` myself. May I?)

**Comment by paulo-ferraz-oliveira** [comment]
> ⚠️ Should `CHANGELOG.md` be updated?
> 
> (I didn't really consider this before, but the TLS-related changes made me think about it - e.g. the consumer upgrades `gen_rpc`, runs into issues because of it [TLS] and doesn't know where to start.
> 
> [This, I guess, is where the thin line between modernizing and _modernizing_ stands.]
> 
> I don't mind updating `CHANGELOG.md` myself. May I?)

Done: I tried to be exhaustive. Let me know if we're missing something.

**Comment by g-andrade** [comment]
Pushed a small tweak on 8c34f99.

**Review by g-andrade — APPROVED** [review]

**Comment by g-andrade** [comment]
I'll leave the merging up to you depending on what you think of 8c34f99

**Comment by paulo-ferraz-oliveira** [comment]
> I'll leave the merging up to you depending on what you think of [8c34f99](https://github.com/miniclip/gen_rpc/commit/8c34f99149d341038ded579c33c94fa4c54f9b2a)

I'm Ok with the tweak. Are you comfortable with the potential breaking changes (.hrl and ciphers) as-is? I mean, I consider these bugfixes, but you might see otherwise.

**Comment by g-andrade** [comment]
> Are you comfortable with the potential breaking changes (.hrl and ciphers) as-is?

I looked around our internal codebase for any uses of the headers (not thoroughly), couldn't find any. So I'm fairly confident it won't break out stuff.

It might break others', but there's only so much we can do to save the world.

**Comment by g-andrade** [comment]
> Are you comfortable with the potential breaking changes (.hrl and ciphers) as-is?

As for the ciphers, yeah it might be an issue in a stack that mixes nodes with both pre- and post-modern `gen_rcp` - _if_ there are no overlapping ciphers for TLS 1.2 (the version they have in common.)

I should have checked that...

**Comment by g-andrade** [comment]
> I should have checked that...

Based on 044fb46:
```erlang
> Old = ["ECDHE-ECDSA-AES256-GCM-SHA384", ...].

>  New = [
             % tlsv1.3
             "TLS_AES_128_GCM_SHA256",
             ...].
             
> New -- (New -- Old).
["ECDHE-ECDSA-AES256-GCM-SHA384",
 "ECDHE-RSA-AES256-GCM-SHA384",
 "ECDHE-ECDSA-AES128-GCM-SHA256",
 "ECDHE-RSA-AES128-GCM-SHA256"]
 ```

So, _assuming_ these 4 ciphers are widely supported, there shouldn't be a problem with mixed clusters.

_Right?_

Hopefully so.

🙄

Oh well.

**Comment by paulo-ferraz-oliveira** [comment]
> New -- (New -- Old)

What's this? `--Old`? 🤣

...

You mean there's old stuff we were supporting we aren't anymore, right? I don't mind keeping 
`tlsv1.1` and phasing it out later, though. If not, we just go 3.0 and state in the Release Notes that ciphers changed and a couple `.hrl`s got removed.

**Comment by g-andrade** [comment]
> What's this? --Old? 🤣

😄 not sure if it's sarcasm or a joke, so I'll interpret it literally: 
- I took any `Old` cipher that remained in `New`, leaving me with "ciphers that are in `New` but not in `Old`"';
- I then took these from `New`, leaving me with "ciphers that are both in `Old` and in `New`" - the ones in common, i.e. the middle ground between the old `gen_rpc` and the new `gen_rpc`.

As for keeping TLS 1.1 around, I'd rather just go `gen_rpc` 3.0.
